File: src/python/problems/min_time_to_visit_all_points_test.py
# https://leetcode.com/problems/minimum-time-visiting-all-points

from core.problem_base import *

class MinTimeToVisitAllPoints(ProblemBase):
    def Solution(self, points: list[list[int]]) -> int:
        result = 0
        for i in range(len(points) - 1):
            result += max(
                abs(points[i][0] - points[i + 1][0]), 
                abs(points[i][1] - points[i + 1][1]))

        return result

if __name__ == '__main__':
    TestGen(MinTimeToVisitAllPoints) \
        .Add(lambda tc: tc.Param([[1,1],[3,4],[-1,0]]).Result(7)) \
        .Add(lambda tc: tc.Param([[3,2],[-2,2]]).Result(5)) \
        .Run()

File: src/go/problems/min_time_to_visit_all_points_test.go
// https://leetcode.com/problems/minimum-time-visiting-all-points

package problems

import (
	"math"
	"testing"

	"github.com/dobermanch/leetcode/core"
)

type MinTimeToVisitAllPoints struct{}

func TestMinTimeToVisitAllPoints(t *testing.T) {
	gen := core.TestSuite[MinTimeToVisitAllPoints]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param(10).Param(13).Result(23)
	}).Add(func(tc *core.TestCase) {
		tc.Param(-10).Param(12).Result(2)
	}).Run(t)
}

func (MinTimeToVisitAllPoints) Solution(points [][]int) int {
	result := 0.0
	for i := 1; i < len(points); i++ {
		result += math.Max(
			math.Abs(float64(points[i-1][0]-points[i][0])),
			math.Abs(float64(points[i-1][1]-points[i][1])))
	}

	return int(result)
}
